pkg/reconciler/eventtype.go

```diff
@@ -0,0 +1,193 @@
+		err = r.Client.Create(ctx, &eventType)
+		if err != nil {
+			return err
+		}
+	}
+	// Deletion might not be needed, but just to make sure we dedupe EventTypes
+	// in case some source controller has a bug.
+	for _, eventType := range sync.ToDelete {
+		err = r.Client.Delete(ctx, &eventType)
+		if err != nil && !errors.IsNotFound(err) {
+			return err
+		}
+	}
+	return nil
+}
+
+// getEventTypes returns the EventTypes controlled by 'owner'.
+func (r *EventTypeReconciler) getEventTypes(ctx context.Context, namespace string, lbs map[string]string, owner metav1.Object) ([]eventingv1alpha1.EventType, error) {
+	eventTypes := make([]eventingv1alpha1.EventType, 0)
+
+	opts := &client.ListOptions{
+		Namespace:     namespace,
+		LabelSelector: labels.SelectorFromSet(lbs),
+		// Set Raw because if we need to get more than one page, then we will put the continue token
+		// into opts.Raw.Continue.
+		Raw: &metav1.ListOptions{},
+	}
+	for {
+		el := &eventingv1alpha1.EventTypeList{}
+		if err := r.Client.List(ctx, opts, el); err != nil {
+			return nil, err
+		}
+
+		for _, e := range el.Items {
+			if metav1.IsControlledBy(&e, owner) {
+				eventTypes = append(eventTypes, e)
+			}
+		}
+		if el.Continue != "" {
+			opts.Raw.Continue = el.Continue
+		} else {
+			return eventTypes, nil
+		}
+	}
+}
+
+// makeEventTypes creates the in-memory representation of the EventTypes.
+func (r *EventTypeReconciler) makeEventTypes(args *EventTypeReconcilerArgs, owner metav1.Object) ([]eventingv1alpha1.EventType, error) {
+	eventTypes := make([]eventingv1alpha1.EventType, 0)
+	for _, spec := range args.EventTypeSpecs {
+		eventType := r.makeEventType(spec, args.Namespace, args.Labels)
+		// Setting the reference to delete the EventType upon uninstalling the source.
+		if err := controllerutil.SetControllerReference(owner, &eventType, r.Scheme); err != nil {
+			return nil, err
+		}
+		eventTypes = append(eventTypes, eventType)
+	}
+	return eventTypes, nil
+}
+
+// makeEventType creates the in-memory representation of the EventType.
+func (r *EventTypeReconciler) makeEventType(spec eventingv1alpha1.EventTypeSpec, namespace string, lbl map[string]string) eventingv1alpha1.EventType {
+	return eventingv1alpha1.EventType{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: fmt.Sprintf("%s-", ToDNS1123Subdomain(spec.Type)),
+			Labels:       lbl,
+			Namespace:    namespace,
+		},
+		Spec: spec,
+	}
+}
+
+// getSyncEventTypes computes the EventTypes that need to be created and/or deleted based on the difference between
+// 'expected' and 'current'. It does so using all the EventType.Spec fields but Description.
+func (r *EventTypeReconciler) getEventTypesToSync(current []eventingv1alpha1.EventType, expected []eventingv1alpha1.EventType) *syncEventTypes {
+	eventTypesToSync := &syncEventTypes{
+		ToCreate: make([]eventingv1alpha1.EventType, 0),
+		ToDelete: make([]eventingv1alpha1.EventType, 0),
+	}
+	keyFunc := func(eventType *eventingv1alpha1.EventType) string {
```

Comment:
What do you think about making this a package function instead? Then it's clear that it doesn't rely on method-local scope. I don't have a strong preference either way.

Reply:
Done, made it a package function instead! Thanks for the comment